src/App.js

```diff
@@ -7,78 +7,65 @@ import { SideBar } from "./components/SideBar";
 import "./App.css";
 
 
-export class App extends Component {
+export const App = React.memo(() => {
 
-	constructor(props) {
-		super(props);
+	const [layers, setLayers] = React.useState([]);
+	const [ready, setReady] = React.useState(false);
+	const driverRef = React.useRef(undefined);
 
-		this.state = {
-			ready: false,
-			layers: []
-		};
-
-		this.saveConfigToFile = this.saveConfigToFile.bind(this);
-		this.loadConfigFromFile = this.loadConfigFromFile.bind(this);
-		this.addLayer = this.addLayer.bind(this);
-		this.updateLayer = this.updateLayer.bind(this);
-		this.removeLayer = this.removeLayer.bind(this);
-	};
+	// This blocks render indefinitely if the driver is never resolved.
+	// A better pattern would be to render anyway without driver,
+	// only calling getDriver() when a module wants to use the driver.
```

Comment:
The problem here is that the Layer component needs data from the graph to render (to get the list of labels or whether neo4j-spatial is available).

Reply:
Is this true when we load data from file?

Reply:
Since we start off with a blank layer that doesn't render on the map, driver comes into play when we create new / update layer. IMO that's when we need to check for the driver, and first layer to check for driver should create an instance usable by any other subsequent ones.

We still get side pane and empty map experience even if we didn't have a driver. At create in case we cant boot up a driver, could we possibly provide a flow for troubleshooting/manually establishing a graph connection within neo4j desktop?

Reply:
True, this does not apply when we load data from file.

But it does appear before the layer create/update, since we not only get data for the map, but also data to configure the layer. In terms of UX, it is important to keep this flow. I suggest that we move this discussion to a new issue though, it is not directly related to this PR.

Reply:
For tracking, discussion about this topic moved to: 
https://github.com/stellasia/neomap/issues/78